compiler/rustc_codegen_ssa/src/mir/naked_asm.rs

```diff
@@ -264,7 +271,132 @@ fn prefix_and_suffix<'tcx>(
                 writeln!(end, "{}", arch_suffix).unwrap();
             }
         }
+        AsmBinaryFormat::Wasm => {
+            let section = link_section.unwrap_or(format!(".text.{asm_name}"));
+
+            writeln!(begin, ".section {section},\"\",@").unwrap();
+            writeln!(begin, ".balign {align}").unwrap();
+            write_linkage(&mut begin).unwrap();
+            if let Visibility::Hidden = item_data.visibility {
+                writeln!(begin, ".hidden {asm_name}").unwrap();
+            }
+            writeln!(begin, ".type {asm_name}, @function").unwrap();
+            if !arch_prefix.is_empty() {
+                writeln!(begin, "{}", arch_prefix).unwrap();
+            }
+            writeln!(begin, "{asm_name}:").unwrap();
+            writeln!(begin, ".functype {asm_name} {}", wasm_functype(tcx, fn_abi)).unwrap();
+
+            writeln!(end).unwrap();
+            // .size is ignored for function symbols, so we can skip it
+            writeln!(end, "end_function").unwrap();
+        }
     }
 
     (begin, end)
 }
+
+/// The webassembly type signature for the given function.
+///
+/// Used by the `.functype` directive on wasm targets.
+fn wasm_functype<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> String {
+    let mut signature = String::with_capacity(64);
+
+    let ptr_type = match tcx.data_layout.pointer_size.bits() {
+        32 => "i32",
+        64 => "i64",
+        other => bug!("wasm pointer size cannot be {other} bits"),
+    };
+
+    let hidden_return =
+        matches!(fn_abi.ret.mode, PassMode::Indirect { .. } | PassMode::Pair { .. });
```

Comment:
On `wasm32-unknown-unknown`, a `#[repr(C)] struct Compound { a: i16, b: i64 }` is returned as a `PassMode::Pair`. But in practice LLVM (?) turns that into a hidden return anyway.

I'm guessing that there are reasons for not using a `PassMode::Indirect` on this target? And this line assumes that a `Pair` is never returned "normally" (i.e. not hidden).

Reply:
For wasm32-unknown-unknown it is kind of expected given that the ABI is broken there. For `extern "C"` with wasm32-wasip1 however we do correctly use `PassMode::Indirect`. As we are going to fix the ABI of wasm32-unknown-unknown to match wasm32-wasip1, I'm not sure a special case here is actually necessary and not adding it may in fact discourage people from depending on the broken ABI.

Reply:
right, so https://github.com/rust-lang/rust/issues/115666. That does mean that this code tries to replicate what LLVM does, which is risky.

What would "not adding" mean? An assert, `bug!`, pretty error message, generate an incorrect signature, or something else?

Reply:
`bug!()` is probably fine. Once the ABI is fixed `PassMode::Pair` in return position should be unreachable on wasm.